[Lang] Struct Classes implementation

bsavery opened the pull request:
Related issue = #4965

### TLDR.  Turns this:
```python
@ti.struct_class
class Sphere:
    center: vec3
    radius: ti.f32

    @ti.func
    def area(self):
        return 4 * 3.14 * self.radius * self.radius
```
Into a taichi struct (with methods!)

As discussed in the issue, this is a (draft) Pull request to implement `struct_classes`.  
`struct_class` is a decorator that can be attached to a python class definition that allows two things:

1.  Class field annotations are turned into struct fields (inspired by python Dataclasses)
2.  Class methods are attached to the struct and used like class methods.

IMO this allows a better syntax for structs and of course allows the struct methods.  Additionally it might ease much of what people would do with `ti.data_oriented` for complex arrays of structures in classes.  

**NOTE THIS IS A DRAFT PR**.  I am looking for feedback of the style and naming.  Also note I copied this from my working version so might need tweaking to work on your side.

### **Some technical notes**. 
There are basically two main things added here:
1.  Struct python type https://github.com/taichi-dev/taichi/blob/09de043b1e5fc522d642e7fb14a9dfed3283c8f3/python/taichi/lang/struct.py#L15 now can handle an optional `__struct_methods` object passed to it which contains the methods attached to the struct type.  I am open to other suggestions how to handle this.  But it should not have much overhead as the struct methods are only kept on the python side.  
2.  The new class decorator that turns a class with annotations fields into a struct and passes the methods.


### Example to test
```python
@ti.struct_class
class Sphere:
    center: vec3
    radius: ti.f32

    @ti.func
    def area(self):
        return 4 * 3.14 * self.radius * self.radius

n = 10
my_spheres = Sphere.field(shape=(n,))
print(my_spheres.radius)

# one struct
my_sphere = Sphere(center=vec3(2.0, 1.0, 9.0), radius=4.0)
# add to field
my_spheres[4] = my_sphere
print('done setup')

@ti.kernel
def kernel():
    sphere = my_spheres[4]
    center = sphere.center
    rad = sphere.radius
    area = sphere.area()
    
kernel()
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/taichi-dev/taichi?pullRequest=4989) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/taichi-dev/taichi?pullRequest=4989) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/taichi-dev/taichi?pullRequest=4989) it.</sub>

strongoier commented on python/taichi/lang/struct.py:
```diff
@@ -49,6 +56,14 @@ def __init__(self, *args, **kwargs):
+        for name, method in self.methods.items():
+            # use MethodType to pass self (this object) to the method
+            setattr(self, name, MethodType(method, self))
+            if getattr(method, '_is_taichi_function', False) is True:
```
I'm a little curious about why we need this. Doesn't `@ti.func` work?

BTW are we supposed to get these struct methods run in both Python scope and Taichi scope?

bsavery replied:
So the fact that we're re-binding the function to a new object with the func() call seems to lose the ti.func bits and gives a warning.  

Python and taichi scope functions should work fine with this change.

strongoier commented on python/taichi/lang/struct.py:
```diff
@@ -632,7 +664,41 @@ def filled_with_scalar(self, value):
+        return Struct.field(self.members, self.methods, **kwargs)
+
+
+def struct_class(cls):
```
I'm not so sure whether the name `ti.struct_class` is easy enough to understand. Is something like `ti.struct_with_methods` a more straightforward choice? @k-ye

k-ye replied:
Hmm, I actually feel like `struct_class` OK enough. Do you think it's not straightforward from the C++ perspective..?

strongoier replied:
To some extent yes :-) Putting `struct` and `class` together confuses me a bit. But it may be OK if we understand it as "turning a Python class into a Taichi struct".

strongoier reviewed (commented):
Thanks for the great work!

bsavery commented on python/taichi/lang/impl.py:
```diff
@@ -210,18 +210,6 @@ def make_tensor_element_expr(_var, _indices, shape, stride):
 
 
-class SrcInfoGuard:
-    def __init__(self, info_stack, info):
```
Yes, I'll get that fixed.

k-ye commented:
Thanks for the PR! I'd like to understand a few things here:

1. Will `ti.struct_class` always AOS?
2. Can we use `ti.field(dtype=StructType, ...)` for this new struct? (We plan to gradually migrate all `ti.Vector.field`, `ti.Matrix.ndarray` to this unified, `dtype` style.)

bsavery commented:
> Thanks for the PR! I'd like to understand a few things here:
> 
> 1. Will `ti.struct_class` always be AOS?
> 2. Can we use `ti.field(dtype=StructType, ...)` for this new struct? (We plan to gradually migrate all `ti.Vector.field`, `ti.Matrix.ndarray` to this unified, `dtype` style.)

No behavior of creating a struct field is changed.  So you can do anything you can with a normal field.

bsavery commented:
@strongoier it looks like the CI tests are failing due to an api change?  Assuming because I'm exposing a new name.  Can you point me where to fix that?

strongoier commented:
> @strongoier it looks like the CI tests are failing due to an api change? Assuming because I'm exposing a new name. Can you point me where to fix that?

Yep. You can fix that in `tests/python/test_api.py`, which restricts all the public APIs.

bsavery commented:
Another point that might be worth pointing out, one could actually do something like this after the change:

```python
def area(self):
    return 4 * 3.14 * self.radius * self.radius
sphere_struct = ti.types.struct(radius=ti.f32, center=vec3, __struct_methods={'area':area})
```

strongoier reviewed (commented):
The technical route looks good to me! I think we can now work towards getting the PR in. Some comments besides in-line reviews:
1. I tried your branch locally and tweaked a bit. It seems that `@ti.func` works fine if you remove the manual marking part (`if getattr(method, '_is_taichi_function', False) is True`). However, there's another problem - `ti.func` can only be called in Taichi scope, but not in Python scope. My proposal here is to let users mark nothing and let `struct_class` mark the function as `pyfunc`: https://github.com/taichi-dev/taichi/blob/ac1f06f36f292d800a0f0b6b2f3b38ea66867fd4/python/taichi/lang/kernel_impl.py#L76
2. Could you add some tests in `tests/python/test_custom_struct.py`? Other tests in that file can serve as examples. We need to make sure the new feature actually works as expected. 
3. The last piece is about docstrings and documentation - we can put this off until all other issues are resolved.

bsavery commented:
> The technical route looks good to me! I think we can now work towards getting the PR in. Some comments besides in-line reviews:
> 
> 1. I tried your branch locally and tweaked a bit. It seems that `@ti.func` works fine if you remove the manual marking part (`if getattr(method, '_is_taichi_function', False) is True`). However, there's another problem - `ti.func` can only be called in Taichi scope, but not in Python scope. My proposal here is to let users mark nothing and let `struct_class` mark the function as `pyfunc`: https://github.com/taichi-dev/taichi/blob/ac1f06f36f292d800a0f0b6b2f3b38ea66867fd4/python/taichi/lang/kernel_impl.py#L76

So I updated the change.  You're right that functions don't need to be re-marked as ti funcs.  I think the user should be able to mark struct class functions as taichi scope or python scope and he can now do that. 
> 2. Could you add some tests in `tests/python/test_custom_struct.py`? Other tests in that file can serve as examples. We need to make sure the new feature actually works as expected.

Added 
> 3. The last piece is about docstrings and documentation - we can put this off until all other issues are resolved.


Let me know what you want here.

strongoier reviewed (commented):
> I think the user should be able to mark struct class functions as taichi scope or python scope and he can now do that.

Sounds good!

> Let me know what you want here.

Suggestions for this PR are commented.

BTW some potential future work comes to my mind, which is not required for this PR: it will be pretty helpful if you can also write a short article/tutorial on our main docsite about using the new style to write Taichi programs, so that this feature can get advertised :-)

bsavery commented:
Will happily write something under the docsite.  Probably a section un "Advanced Topics"?

strongoier commented on python/taichi/lang/struct.py:
```diff
@@ -233,24 +244,29 @@ def __str__(self):
         Args:
-
+            include_methods (bool): Whether any struct methods should be included
+                in the result dictionary under the key '__struct_methods'.
```
```suggestion
                in the result dictionary under the key '__struct_methods'.

```

strongoier commented on python/taichi/lang/struct.py:
```diff
@@ -336,17 +355,21 @@ def field(cls,
 
     Args:
         entries (Dict[str, Union[Expr, Matrix, Struct]]): keys and values for struct members.
+            Any methods included under the key '__struct_methods' will be applied to each
```
This should not be a separate paragraph.

strongoier commented on python/taichi/lang/struct.py:
```diff
@@ -629,10 +660,53 @@ def filled_with_scalar(self, value):
 
-__all__ = ["Struct", "StructField"]
+def struct_class(cls):
+    """ Converts a class with field annotations into a struct type for use as a taichi field
```
```suggestion
    """Converts a class with field annotations and methods into a taichi struct type.
```

strongoier commented on python/taichi/lang/struct.py:
```diff
@@ -629,10 +660,53 @@ def filled_with_scalar(self, value):
+        >>>     @ti.func
+        >>>     def area(self):
+        >>>         return 4 * 3.14 * self.radius * self.radius
+        >>> my_spheres = Sphere.field(shape=(n,))
```
```suggestion
        >>>
        >>> my_spheres = Sphere.field(shape=(n, ))
```

strongoier reviewed (approved):
LGTM!!! Only some nits left. Thanks again for the efforts!

> Will happily write something under the docsite. Probably a section un "Advanced Topics"?

Sounds great! FYI the docs are under `taichi/docs/lang/articles/advanced/`.

bsavery commented:
Awesome!  I’ll write some documentation this week asap
